### ISSUE TITLE
List<T>.Clear() is surprisingly expensive in timing test.

### ISSUE THREAD

**Issue by dbh1997** [sq/JSIL#762]
It's approx 200 times slower than C#.  This is with a completely empty list,  In fact it takes less time to reconstruct the class (about twice as fast).  Pretest is helpful in either case (Raw C#/JSIL) where the list might already be empty...   I really would have expected this to be the first thing that the IL code does, but apparently not.  I will attempt to test with elements as well to see what sort of difference it makes - I will have to isolate out the costs of the Add() calls to be accurate.  

Might be a really good reason for the performance on this one, but thought I would report it in case it's something that can be corrected.

List<T>.Clear() empty list
........................................C# Time 274 MS, JS Time 58498 MS (213.5x)
List<T>.Clear() empty list with pre-test
........................................C# Time 69 MS, JS Time 347 MS (5.0x): 99.4% Decrease in Time
List<T> reconstruct empty list
........................................C# Time 1659 MS, JS Time 26600 MS (16.0x): 54.5% Decrease in Time

        public static void ClearListEmpty()
        {
            Console.WriteLine("{0}", new String('=', 80));
            const int loopCount = 100000000;

            int test1Time, test2Time, test3Time;

            List<int> list = new List<int>();

            {
                var cSharpTime = 274;
                var start = GetMilliseconds();

                for (int i = 0; i < loopCount; i++)
                {
                    list.Clear();
                }
                var end = GetMilliseconds();
                var runTime = end - start;
                OutputTimes("List<T>.Clear() empty list", cSharpTime, runTime);
                test1Time = runTime;
            }

            {
                var cSharpTime = 69;
                var start = GetMilliseconds();

                for (int i = 0; i < loopCount; i++)
                {
                    if (list.Count > 0)
                        list.Clear();
                }
                var end = GetMilliseconds();
                var runTime = end - start;
                OutputTimes("List<T>.Clear() empty list with pre-test", cSharpTime, test1Time, runTime);
                test2Time = runTime;
            }

            {
                var cSharpTime = 1659;
                var start = GetMilliseconds();

                for (int i = 0; i < loopCount; i++)
                {
                    list = new List<int>();
                }
                var end = GetMilliseconds();
                var runTime = end - start;
                OutputTimes("List<T> reconstruct empty list", cSharpTime, test1Time, runTime);
                test3Time = runTime;
            }
        }

**Comment by kg** [comment]
Welcome to JavaScript. :-( This is because the items have to be zeroed out so this is done by setting the backing store's length to 0 and then back to the correct value.

It might be better to use JSIL's Array.Clear implementation there and see if it's faster. It might be.

**Comment by dbh1997** [comment]
Negligible time saving on ArrayList.Clear() - well within normal variance.  Would be more costly and less readable with all the casts - and boxing/unboxing. 

Looks like time goes up modesty as items are added as expected.  You take a large hit for calling the Clear function regardless then a little more for each element after that..  Firefox does do a better job, but it still pretty slow (around 80x).  

Next I tried replacing clear with a loop that manually deleted the items moving backwards.  It's substantially more expensive than calling Clear() - which would have been surprising if it wasn't.

Curiously, even .NET executes the constructor faster than the Clear() once items are added.  As long as you can specify the initial length as much as you plan on adding, the overall loop is faster.  You probably only gain if you can keep it to the very short lived garbage collection set too, so probably not a recommended thing to do in general - but interesting nevertheless.

Thanks again for all you do!

**Comment by kg** [comment]
Depending on the type of the items, you can optimize out the array clear. It's only necessary for reference types or structs that can contain references (for correctness, to ensure that an empty/shrunk list doesn't retain arbitrary GC objects forever). That might solve your issue here, but it's a hairy problem.

**Comment by dbh1997** [comment]
Good point, I will have to try it with a class as well.  List<int> is pretty popular as an ID collection in our code but the other is used as well

Just to add some complication.  One more test I didn't consider.  list.RemoveRange(0, list.Count).  About 10% slower in .NET than Clear(), but substantially faster in JSIL - about 3x.  Haven't tested it for correctness though to be honest...

Was unable to test the 0 case.  .NET allows the count to be 0, JSIL throws an exception of 0 is passed in (so another minor defect I reckon).  Seems that it could be optimized to return immediately on a zero (as could Clear if Count is 0).
